### DISCUSSION TITLE
Storing a value and repeating it?

### DISCUSSION THREAD

**Discussion by lynxmeninx** [ImageMagick/ImageMagick#5370] [Help] [answered]
I am so sorry for asking what I'm certain is not only a really moronic question, but also one that isn't related specifically to ImageMagick -- but I am having a really hard time both wording and understanding the solution to my request here, and I was hoping someone here might be able to decipher what I'm looking for! This is the first time I've ever done any sort of work in the command line and I have **absolutely** no idea what I'm doing. As before, I'm using ImageMagick 7.1.0 Q16 (windows x64).

(with the help of the lovely people who have answered my other questions here!!,) I now have a code that does a couple of things: 1. It makes a sprite sheet out of a few gifs, collates them exactly how I need them to be, and 2. also creates a thumbnail out of the first frame of the first gif.
```
magick -background none ^
   ( stand*right.gif -coalesce -duplicate 20,0 -gravity center -extent 90x90 +append ) ^
   ( walk*right.gif -coalesce -gravity center -extent 90x90 +append ) ^
-gravity northwest -append -filter point -resize 50% -extent 1080x1080 -write spritesheet.png +delete ^
   ( stand*right.gif -coalesce -delete 1-100 -gravity center -extent 90x90 +append ) ^
-filter point -resize 50% thumb.png
```

However, as you can see, I need to repeat `-extent 90x90` a few times in this one - and this is the most simple spritesheet out of the few hundred I have to do this for! All of them are slightly different sizes, and hence the extent value needs to be slightly different, which means I need to replace 90x90 with the respective value in every line, every time I run this code.

Is there a way that I can tell the command line to hold onto the value that I need for every line just once, so that I only have to change it one time every time I run it? I've had a look at -define, -set, and a couple of other things, and they all flew over my head because I didn't recognize this specific use-case in there. If anyone here would be able to explain it to me as if I were five I would appreciate that enormously. That's about where I feel intellectually.

Secondary to that question - is it possible to do the same thing for a portion of the file names of the outputs? As in, I'd ask for it to recognize [charactername] and replace it with bob at the start, and have `[charactername]spritesheet.png` and `[charactername]thumbnail.png` as my output names, then having the actual file names be bobspritesheet.png and bobthumbnail.png, without having to touch those lines of code? 

Thank you so much -- I'm a little worried this question is overly general, but wanted to ask here just in case anything changes when using ImageMagick.

**Comment by snibgo** [comment, marked as answer]
> Is there a way that I can tell the command line to hold onto the value that I need for every line just once, so that I only have to change it one time every time I run it?

One obvious solution is to use an environment variable (aka shell variable). In Windows, we can set a variable like this:
```
set EXT=-extent 90x90
```
My variable is named `EXT`. (Names don't need to be in CAPITALS, but I usually do, for legibility within the IM command.) When I use it, I need to put it within percent signs: `%EXT%`. When the shell interpreter sees this, it will substitute the current value.
```
magick -background none ^
   ( stand*right.gif -coalesce -duplicate 20,0 -gravity center %EXT% +append ) ^
   ( walk*right.gif -coalesce -gravity center -%EXT% +append ) ^
-gravity northwest -append -filter point -resize 50% -extent 1080x1080 -write spritesheet.png +delete ^
   ( stand*right.gif -coalesce -delete 1-100 -gravity center %EXT% +append ) ^
-filter point -resize 50% thumb.png
```
If you do this within BAT scripts, don't double the percent signs around environment variables.

> Secondary to that question - is it possible to do the same thing for a portion of the file names of the outputs?

Yes. If "bob" is the character-name part of filenames, you can:
```
set CHARNAME=bob

magick -background none ^
   ( stand%CHARNAME%right.gif -coalesce -duplicate 20,0 -gravity center %EXT% +append ) ^
   ( walk%CHARNAME%right.gif -coalesce -gravity center -%EXT% +append ) ^
-gravity northwest -append -filter point -resize 50% -extent 1080x1080 -write %CHARNAME%spritesheet.png +delete ^
   ( stand*right.gif -coalesce -delete 1-100 -gravity center %EXT% +append ) ^
-filter point -resize 50% %CHARNAME%thumb.png
```
